src/webgpu/api/operation/shader_module/compilation_info.spec.ts

```diff
@@ -1,156 +1,184 @@
-    for (const message of info.messages) {
-      if (message.type == 'error') {
-        // Some backends may not be able to indicate a precise location for the error. In those
-        // cases a line and position of 0 should be reported.
-        if (message.lineNum == 0) {
-          foundAppropriateError = true;
-          t.expect(message.linePos == 0, "GPUCompilationMessages that don't report a line number should not report a line position.")
-          break;
-        }
-
-        // If a line is reported, it should point at the correct line (1-based).
-        if (message.lineNum == 4) {
-          foundAppropriateError = true;
-
-          // Various backends may choose to report the error at different positions within the line,
-          // so it's difficult to meaningfully validate them.
-          break;
-        }
-      }
-    }
-    t.expect(foundAppropriateError, "Expected to find an error which corresponded with the erronious line");
-});
-
-
-g.test('compilationInfo_offset_length')
-  .desc(
-    `Ensures message offsets and lengths are valid and align with any reported lineNum and linePos.`
-  )
-  .fn(async t => {
-
-    const shaderModule = t.expectGPUError(
-      'validation', // Should be a validation error since the shader is invalid.
-      () =>
-        t.device.createShaderModule({
-          code: BAD_SHADER_SOURCE,
-        })
-    );
-
-    const info = await shaderModule.compilationInfo();
-
-    for (const message of info.messages) {
-      // Any offsets and lengths should reference valid spans of the shader code.
-      t.expect(message.offset < BAD_SHADER_SOURCE.length, "Message offset should be within the shader source");
-      t.expect(message.offset + message.length < BAD_SHADER_SOURCE.length, "Message offset and length should be within the shader source");
-
-      // If a valid line number and position are given the offset should point the the same location in the shader source.
-      if (message.lineNum != 0 && message.linePos != 0) {
-        let lineOffset = 0;
-        for (let i = 0; i < message.lineNum-1; ++i) {
-          lineOffset = BAD_SHADER_SOURCE.indexOf('\n', lineOffset);
-          assert(lineOffset != -1);
-          lineOffset+=1;
-        }
-
-        t.expect(message.offset == lineOffset + message.linePos-1, "lineNum and linePos should point to the same location as offset");
-      }
-    }
-});
-
-g.test('compilationInfo_offset_unicode')
-  .desc(
-    `Ensures that message offsets properly take into account unicode characters`
-  )
-  .unimplemented();
+export const description = `
+ShaderModule CompilationInfo tests.
+`;
+
+import { makeTestGroup } from '../../../../common/framework/test_group.js';
+import { assert } from '../../../../common/framework/util/util.js';
+import { GPUTest } from '../../../gpu_test.js';
+
+export const g = makeTestGroup(GPUTest);
+
+const kValidShaderSources = [
+  {
+    valid: true,
+    unicode: false,
+    codeFn: () => `
```

Comment:
You can start a parameter name with `_` to make it "private" (not a part of the "identity" of the case). Here I think you can replace the function with the string, i.e. ``_code: `...` ``

(All of the parameters in this list _could_ be private, except that then the cases would not have distinguishable IDs. Using valid+unicode as the ID makes sense.)

Reply:
Oh, good! That's why I ended up going with the function in the first place, so I'm glad to see there's a better solution in place for it.